les/clientpool.go

```diff
@@ -0,0 +1,599 @@
+func (f *clientPool) setClientPriceFactors(c *clientInfo) {
+	c.balanceTracker.setFactors(true, f.negFactors.timeFactor+float64(c.capacity)*f.negFactors.capacityFactor/1000000, f.negFactors.requestFactor)
+	c.balanceTracker.setFactors(false, f.posFactors.timeFactor+float64(c.capacity)*f.posFactors.capacityFactor/1000000, f.posFactors.requestFactor)
+}
+
+// clientPoolStorage is the RLP representation of the pool's database storage
+type clientPoolStorage struct {
+	LogOffset uint64
+	List      []*negBalance
+}
+
+// loadFromDb restores pool status from the database storage
+// (automatically called at initialization)
+func (f *clientPool) loadFromDb() {
+	enc, err := f.db.Get([]byte("clientPool"))
+	if err != nil {
+		return
+	}
+	var storage clientPoolStorage
+	err = rlp.DecodeBytes(enc, &storage)
+	if err != nil {
+		log.Error("Failed to decode client list", "err", err)
+		return
+	}
+	f.logOffsetAtStartup = int64(storage.LogOffset)
+	f.startupTime = f.clock.Now()
+	for _, e := range storage.List {
+		log.Debug("Loaded free client record", "address", e.address, "logValue", e.logValue)
+		f.negBalanceMap[e.address] = e
+		f.negBalanceQueue.Push(e, -e.logValue)
+	}
+}
+
+// saveToDb saves pool status to the database storage
+// (automatically called during shutdown)
+func (f *clientPool) saveToDb() {
+	now := f.clock.Now()
+	storage := clientPoolStorage{
+		LogOffset: uint64(f.logOffset(now)),
+	}
+	for _, c := range f.connectedMap {
+		f.finalizeBalance(c, now)
+	}
+	i := 0
+	storage.List = make([]*negBalance, len(f.negBalanceMap))
+	for _, e := range f.negBalanceMap {
+		storage.List[i] = e
+		i++
+	}
+	enc, err := rlp.EncodeToBytes(storage)
+	if err != nil {
+		log.Error("Failed to encode negative balance list", "err", err)
+	} else {
+		f.db.Put([]byte("clientPool"), enc)
+	}
+	for _, b := range f.posBalanceMap {
+		f.storePosBalance(b)
+	}
+}
+
+// storePosBalance stores a single positive balance entry in the database
+func (f *clientPool) storePosBalance(b *posBalance) {
+	if b.value == b.lastStored {
+		return
+	}
+	enc, err := rlp.EncodeToBytes(b)
+	if err != nil {
+		log.Error("Failed to encode client balance", "err", err)
+	} else {
+		f.db.Put(append([]byte("clientPool-balance"), b.id[:]...), enc)
+	}
+}
+
+// getPosBalance retrieves a single positive balance entry from cache or the database
+func (f *clientPool) getPosBalance(id enode.ID) *posBalance {
+	if b, ok := f.posBalanceMap[id]; ok {
+		f.posBalanceQueue.Remove(b.queueIndex)
+		f.posBalanceAccessCounter--
+		f.posBalanceQueue.Push(b, f.posBalanceAccessCounter)
+		return b
+	}
+	balance := &posBalance{}
+	if enc, err := f.db.Get(append([]byte("clientPool-balance"), id[:]...)); err == nil {
+		if err := rlp.DecodeBytes(enc, balance); err != nil {
+			log.Error("Failed to decode client balance", "err", err)
+			balance = &posBalance{}
+		}
+	}
+	balance.id = id
+	balance.queueIndex = -1
+	if f.posBalanceQueue.Size() >= f.queueLimit {
```

Comment:
posBalanceQueue is an LRU. Maybe I could use another implementation, it just felt convenient because I used Prque-s everywhere. Caching non-existent entries is correct because trying to read a non-existent db entry has a cost too.